### PR TITLE
Implement Boards main logic

### PR DESCRIPTION
**Pull request by IraVolyk**
#8 

1. Added Bord model
2. Added ability to create/edit/destroy the board
3. Added search by title and board description
4. Added migration for the composite index (title, description)
5. Added filter by type of boards public/private/user_bords
6. Added views
7. Added gem 'haml-rails'

### CODE REVIEW

**Comment by Synkevych** [comment]
Why did you use an extra gem **erb2haml** to convert from erb to haml ? **haml-rails** can do it for you: 
`$ rails haml:erb2haml`

**Comment by IraVolyk** [comment]
> Why did you use an extra gem **erb2haml** to convert from erb to haml ? **haml-rails** can do it for you:
> `$ rails haml:erb2haml`

I agree we need to uninstall gem erb2haml

**Comment by Synkevych** [comment]
Also set **rails-halm** to the development section, I think they're not needed in production.

https://stackoverflow.com/questions/99211/how-do-i-get-haml-to-work-with-rails

File: app/models/user.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 class User < ApplicationRecord
-  # Include default devise modules. Others available are:
-  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  has_many :boards
```
Comment (yuravenzhik) [inline app/models/user.rb:2]: dependent: :nullify

File: config/routes.rb [hunk 2]
```diff
@@ -1,4 +1,9 @@
 Rails.application.routes.draw do
+
+  get "boards/filter/:filter/" => "boards#index", as: :filtered_boards
```
Comment (yuravenzhik) [inline config/routes.rb:3]: !

File: config/routes.rb [hunk 3]
```diff
@@ -1,4 +1,9 @@
...
+
+  resources :users
```
Comment (yuravenzhik) [inline config/routes.rb:5]: !

File: db/schema.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+ActiveRecord::Schema.define(version: 2020_04_26_160435) do
+
+  # These are extensions that must be enabled in order to support this database
+  enable_extension "plpgsql"
+
+  create_table "boards", force: :cascade do |t|
+    t.string "title"
+    t.text "description"
+    t.boolean "public", default: false
```
Comment (yuravenzhik) [inline db/schema.rb:21]: !

File: app/controllers/boards_controller.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
+class BoardsController < ApplicationController
+	before_action :find_board, only: [:show, :edit,:update, :destroy]
+
+  def index 
+    if params.has_key? :filter
```
[the remark this answers is not in the document]
Comment (etherell) [reply app/controllers/boards_controller.rb:5]: Сделано

File: app/controllers/boards_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  def index 
+    if params.has_key? :filter 
+      @filter = params[:filter]
+      @boards = Board.filter(@filter).search(params[:search])
+    else
+      @boards = Board.all.search(params[:search])
+    end
+
+    @private_boards = @boards.private_boards
+    @public_boards = @boards.public_boards
```
[the remark this answers is not in the document]
Comment (etherell) [reply app/controllers/boards_controller.rb:13]: Сделано

File: app/controllers/boards_controller.rb [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  end
+
+private
+
+  def find_board
+    @board = Board.find(params[:id])
+  end
+
+  def board_params
+    params.require(:board).permit(:title, :description, :search, :public)
```
[the remark this answers is not in the document]
Comment (etherell) [reply app/controllers/boards_controller.rb:62]: Сделано

File: config/routes.rb (hunk 2 above)
Comment (etherell) [reply config/routes.rb:3]: Сделано, удалил роуты

File: config/routes.rb (hunk 3 above)
Comment (etherell) [reply config/routes.rb:5]: Удалили роуты для юзера из нашей ветки

File: db/schema.rb (hunk 4 above)
Comment (IraVolyk) [reply db/schema.rb:21]: Index added

File: app/models/user.rb (hunk 1 above)
Comment (IraVolyk) [reply app/models/user.rb:2]: Added

File: app/controllers/boards_controller.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @board = Board.new
+  end
+
+  def edit
+  end
+
+
+  def create
+    @board = Board.new(board_params)
+    @board.user_id = current_user.id
```
Comment (yuravenzhik) [inline app/controllers/boards_controller.rb:22]: 1

File: app/models/board.rb [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  validates :description, length: { minimum: 5 }
+  validates :user_id, presence: true    
+
+  scope :search, ->(input) { where("title ilike ? or description ilike ?", "%#{input}%", "%#{input}%") }
+  scope :user_boards, ->(user) { where(user_id: user.id) }
+  scope :public_boards, -> { where(public: true) }
+  scope :private_boards, -> { where(public: false) }
+  scope :pagination, ->(page_params) { paginate(page: page_params, per_page: 10).order('created_at DESC') }
+  
+
```
Comment (yuravenzhik) [inline app/models/board.rb:14]: !

File: app/models/board.rb [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  validates :title, length: { minimum: 5 }
+  validates :description, length: { minimum: 5 }
+  validates :user_id, presence: true    
+
+  scope :search, ->(input) { where("title ilike ? or description ilike ?", "%#{input}%", "%#{input}%") }
+  scope :user_boards, ->(user) { where(user_id: user.id) }
+  scope :public_boards, -> { where(public: true) }
+  scope :private_boards, -> { where(public: false) }
+  scope :pagination, ->(page_params) { paginate(page: page_params, per_page: 10).order('created_at DESC') }
```
Comment (yuravenzhik) [inline app/models/board.rb:12]: !

File: app/views/boards/_board.html.haml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+- boards.each do |board|
+	
+	%p
+		%strong title:
```
Comment (yuravenzhik) [inline app/views/boards/_board.html.haml:4]: !

File: app/models/board.rb [hunk 12]
```diff
@@ -0,0 +1,27 @@
+class Board < ApplicationRecord
+  belongs_to :user 
+
+  validates :title, length: { minimum: 5 }
```
Comment (yuravenzhik) [inline app/models/board.rb:4]: !

File: app/models/board.rb [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  validates :description, length: { minimum: 5 }
+  validates :user_id, presence: true
```
Comment (yuravenzhik) [inline app/models/board.rb:6]: !

File: app/controllers/boards_controller.rb [hunk 14]
```diff
@@ -0,0 +1,57 @@
+class BoardsController < ApplicationController
+	before_action :find_board, only: [:show, :edit,:update, :destroy]
```
Comment (yuravenzhik) [inline app/controllers/boards_controller.rb:2]: !

File: app/controllers/boards_controller.rb (hunk 8 above)
Comment (IraVolyk) [reply app/controllers/boards_controller.rb:22]: fix

File: app/models/board.rb (hunk 9 above)
Comment (IraVolyk) [reply app/models/board.rb:14]: done

File: app/models/board.rb (hunk 10 above)
Comment (IraVolyk) [reply app/models/board.rb:12]: deleted

File: app/controllers/boards_controller.rb (hunk 14 above)
Comment (IraVolyk) [reply app/controllers/boards_controller.rb:2]: ! added

File: app/models/board.rb (hunk 12 above)
Comment (IraVolyk) [reply app/models/board.rb:4]: max length added

File: app/models/board.rb (hunk 13 above)
Comment (IraVolyk) [reply app/models/board.rb:6]: add max length and change to validates :user

File: app/views/boards/_board.html.haml (hunk 11 above)
Comment (IraVolyk) [reply app/views/boards/_board.html.haml:4]: corrected

File: app/controllers/boards_controller.rb [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    @board = current_user.boards.new(board_params)
+    if @board.save
+      redirect_to @board
+    else
+      render 'new'
+    end
+  end
+
+  def update
+    flash[:success] = "Successfully updated!"
```
Comment (yuravenzhik) [inline app/controllers/boards_controller.rb:30]: why is it placed here?

**Pull request merged**